### PR TITLE
Add mathjax block and mathjax chemistry

### LINKED ISSUE #18174: MathJax block and chemistry in note editor
Anki desktop offers to add mathjax block and mathjax chemistry. I believe ankidroid should offer the same option.
However, the MathJax button default behavior should not be changed. I expect the mathJax inline to remain the default. 
I believe the most natural way to offer the option of the alternative is to request the user to long tap on the Sigma_\oplus symbol, showing a menu that show the three options. 

I would like to note that I think we should not offer the LaTeX option. As ankidroid can't compile latex, this would just be too confusing.

### PR DESCRIPTION
**Pull request by alisherbu**
## Purpose / Description
Adding mathjax block and mathjax chemistry options by showing a menu dialog

## Fixes
* Fixes #18174 

## Approach
_How does this change address the problem?_

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration (SDK version(s), emulator or physical, etc)

## Learning (optional, can help others)
_Describe the research stage_

_Links to blog posts, patterns, libraries or addons used to solve this problem_

## Checklist
_Please, go through these checks before submitting the PR._

- [ ] You have a descriptive commit message with a short title (first line, max 50 chars).
- [ ] You have commented your code, particularly in hard-to-understand areas
- [ ] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Comment by DrunkenCloud** [comment]
Looks good! It could be nice to change 'MathJax chemistry' to 'MathJax Chemistry' (with a capital C) for consistency, same for others, but works as intended in the issue.

**Comment by alisherbu** [comment]
@DrunkenCloud I had tried to follow desktop. There was lower case. 
image

**Comment by DrunkenCloud** [comment]
Oh nvm mb then

File: AnkiDroid/src/main/res/values/constants.xml [hunk 1]
```diff
@@ -334,4 +334,10 @@
         <item>android.permission.WRITE_EXTERNAL_STORAGE</item>
     </string-array>
 
+    <string-array name="mathjax_equations">
+        <item>MathJax inline</item>
```
Comment (lukstbit) [inline AnkiDroid/src/main/res/values/constants.xml:338]: These should be translatable so each should be declared as a string reference. See some of the arrays above how it's done. I would also name the array as _mathjax_type_.

File: AnkiDroid/src/main/java/com/ichi2/anki/noteeditor/Toolbar.kt [hunk 2]
```diff
@@ -306,6 +310,23 @@ class Toolbar : FrameLayout {
         }
     }
 
+    /**
+     * Displays a dialog that allows the user to insert a MathJax equation in different formats.
+     */
+    @SuppressLint("CheckResult")
+    private fun displayInsertMathJaxEquationsDialog() {
+        val equations = resources.getStringArray(R.array.mathjax_equations)
+        val prefixes = arrayOf("\\(", "\\[\\", "\\[ \\ce{")
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/anki/noteeditor/Toolbar.kt:319]: The first prefix/suffix pair is already inserted when just clicking the button. So only the two other options should be presented on a button long click.

**Review by lukstbit — CHANGES_REQUESTED** [review]
Thank you for contributing, I left some issues to fix.

File: AnkiDroid/src/main/res/values/constants.xml (hunk 1 above)
Comment (david-allison) [reply AnkiDroid/src/main/res/values/constants.xml:338]: These should pull the strings directly from the Desktop:

- `TR.editingMathjaxInline()`
- `TR.editingMathjaxBlock()`
- `TR.editingMathjaxChemistry()`

File: AnkiDroid/src/main/java/com/ichi2/anki/noteeditor/Toolbar.kt (hunk 2 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/noteeditor/Toolbar.kt]: I'd be curious to have a third opinion.
I feel like the first option does not hurt. If the user long pressed by accident, it's better not to have to close the view. Given that this is what I requested in the issue I opened, I must note at least that @alisherbu correctly implemented my request here.

One thing that I had not realized. On most button, long press show a tool tip explaining the action caused by the short press. Given that this feature disable the tooltip, I expect that we should still have a way to let the user know what `\(` and `\)` corresponds too. Having this first option seems the less bad way to do it.

**Comment by Arthur-Milchior** [comment]
One thing that would be really helpful, but not mandatory, is a short video capture of you using your feature. That would allow me to do a first safety check without having to run the code.

File: AnkiDroid/src/main/java/com/ichi2/anki/noteeditor/Toolbar.kt [hunk 3]
```diff
@@ -306,6 +310,23 @@ class Toolbar : FrameLayout {
...
+        val suffixes = arrayOf("\\)", "\\]", "} \\]")
+        AlertDialog.Builder(context).show {
+            setItems(equations) { _, index ->
+                val formatter = TextWrapper(prefix = prefixes[index], suffix = suffixes[index])
+                onFormat(formatter)
+            }
+            title(R.string.insert_mathjax)
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/noteeditor/Toolbar.kt:327]: This string is "Insert mathjax equation". Which was surprising to me, given that it allows to enter chemistry symbols. But I've discovered by looking at https://docs.mathjax.org/en/latest/input/tex/extensions/mhchem.html that it's actually called "Chemical equation". I'd love either your commit message, or this line of code, to have a comment that explains to non chemist that it's called "chemical equation" and so this title is acceptable. That was not intuitive to me, and I'm certain I may not be the only coder who have forgotten everything about chemistry

**Review by Arthur-Milchior — COMMENTED** [review]
Really great first application.
I'd appreciate if you could have a single commit (amend your commit instead of adding a new one) and have a more descriptive description.

The commit message could contain "Fixes #18174". It could also state explain what was in the bug, so that any future reader of our commit history can know the context without opening github. That is, that we follow desktop feature. Except that we don't offer LaTeX because it can't be built on phone, so that would be a frustrating experience

File: AnkiDroid/src/main/java/com/ichi2/anki/noteeditor/Toolbar.kt [hunk 4]
```diff
@@ -306,6 +311,23 @@ class Toolbar : FrameLayout {
...
+     * Displays a dialog that allows the user to insert a MathJax equation in different formats.
+     */
+    @SuppressLint("CheckResult")
+    private fun displayInsertMathJaxEquationsDialog() {
+        val equations = arrayOf(TR.editingMathjaxBlock(), TR.editingMathjaxChemistry())
+        val prefixes = arrayOf("\\[\\", "\\[ \\ce{")
+        val suffixes = arrayOf("\\]", "} \\]")
+        AlertDialog.Builder(context).show {
+            setItems(equations) { _, index ->
+                val formatter = TextWrapper(prefix = prefixes[index], suffix = suffixes[index])
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/noteeditor/Toolbar.kt:324]: I'm on my phone, but don't see how this works: you have 3 elements in the labels, but only 2 prefixes/suffixes.

I feel this would be easier to understand if you had a list of triples, rather than 3 lists of data

Comment (alisherbu) [reply AnkiDroid/src/main/java/com/ichi2/anki/noteeditor/Toolbar.kt:323]: Using Triple improves readability, but it comes at a performance cost. Here's why:
- Triple is a data class, meaning each instance creates an extra object.
- When using arrayOf(Triple(), Triple()), you need to map it to a list of strings and then convert it back to an array to set AlertDialog.setItems() because setItems() method accepts an array. This results in unnecessary List and Array creation.

File: AnkiDroid/src/main/java/com/ichi2/anki/noteeditor/Toolbar.kt [hunk 5]
```diff
@@ -306,6 +310,23 @@ class Toolbar : FrameLayout {
         }
     }
 
+    /**
+     * Displays a dialog that allows the user to insert a MathJax equation in different formats.
+     */
+    @SuppressLint("CheckResult")
```
[the remark this answers is not in the document]
Comment (alisherbu) [reply AnkiDroid/src/main/java/com/ichi2/anki/noteeditor/Toolbar.kt]: Thanks for catching it! It's my bad. I removed it

**Comment by alisherbu** [comment]
Here is the demonstration

https://github.com/user-attachments/assets/c8a41db4-60e0-4345-9929-954b1b9ea489

File: AnkiDroid/src/main/java/com/ichi2/anki/noteeditor/Toolbar.kt (hunk 4 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/noteeditor/Toolbar.kt:323]: Code readability, maintainability and removal of a class of bugs is much more important. 

This would be optimizing for microseconds at worst, probably nanoseconds.

**Review by david-allison — APPROVED** [review]
Needs a readability fix, then LGTM (patch untested)

Thanks so much!

```patch
Index: AnkiDroid/src/main/java/com/ichi2/anki/noteeditor/Toolbar.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/noteeditor/Toolbar.kt b/AnkiDroid/src/main/java/com/ichi2/anki/noteeditor/Toolbar.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/noteeditor/Toolbar.kt	(revision a370c2dd1d383a926ca30ad8b3a7b0a1ec4046f3)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/noteeditor/Toolbar.kt	(date 1743408115636)
@@ -315,13 +315,18 @@
      * Displays a dialog that allows the user to insert a MathJax equation in different formats.
      */
     private fun displayInsertMathJaxEquationsDialog() {
-        val equations = arrayOf(TR.editingMathjaxBlock(), TR.editingMathjaxChemistry())
-        val prefixes = arrayOf("\\[\\", "\\( \\ce{")
-        val suffixes = arrayOf("\\]", "} \\)")
+        data class MathJaxOption(val label: String, val prefix: String, val suffix: String) {
+            fun toTextWrapper() = TextWrapper(prefix = this.prefix, suffix = this.suffix)
+        }
+
+        val mathjaxOptions = arrayOf(
+            MathJaxOption(TR.editingMathjaxBlock(), "\\[\\", "\\]"),
+            MathJaxOption(TR.editingMathjaxChemistry(), "\\( \\ce{", "} \\)")
+        )
+
         AlertDialog.Builder(context).show {
-            setItems(equations) { _, index ->
-                val formatter = TextWrapper(prefix = prefixes[index], suffix = suffixes[index])
-                onFormat(formatter)
+            setItems(mathjaxOptions.map { it.label }.toTypedArray()) { _, index ->
+                onFormat(mathjaxOptions[index].toTextWrapper())
             }
             title(R.string.insert_mathjax)
         }
```

**Comment by alisherbu** [comment]
The feedback from @david-allison has been made! Thanks for explaing the importance of code readability, maintainability.
NIT: There is some unwanted code formatting issues by ktlint

**Review by lukstbit — APPROVED** [review]
LGTM, thanks!

@Arthur-Milchior left a request for a comment/commit message related to the dialog title that was not implemented. There was also the discussion about including the default option in the dialog, I'm fine with the current implementation, so leaving it to @Arthur-Milchior.

**Review by mikehardy — APPROVED** [review]
This is fine as is IMHO, the string Arthur is talking about isn't that non-intuitive and at a code level where developers will mostly work it just says `insert_mathjax` which is exactly what it is doing, I'm queueing this

**Comment by david-allison** [comment]
@alisherbu FYI: Your git email does not match your GitHub account, so your contributions may not appear as a 'pip' on the GitHub heatmap

https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-email-preferences/setting-your-commit-email-address

I'm planning to credit you under your GitHub name for our release, but please let me know if you're anonymous by choice.

**Comment by alisherbu** [comment]
@david-allison Thanks for the heads-up! I’m not intentionally anonymous — my connected GitHub email should be alisherqazaqbaev@gmail.com

**Comment by ZornHadNoChoice** [comment]
Hi thanks for the feature, just a minor nitpick: the inserted text should be `\[\]` without the extra backslash.

**Comment by david-allison** [comment]
Hi there @alisherbu! This is the OpenCollective Notice for PRs merged from [2025-04-01 through 2025-04-30](https://github.com/ankidroid/Anki-Android/pulls?q=is%3Apr+merged%3A2025-04-01..2025-04-30+author%3Aalisherbu)
                      
If you are interested in compensation for this work, the process with details is here: https://github.com/ankidroid/Anki-Android/wiki/OpenCollective-Payment-Process#how-to-get-paid

We only post one comment per person per month to avoid spamming you, regardless of the number of PRs merged, but this note applies to all PRs merged for this month

Please understand that our monthly budget is never guaranteed to cover all claims - the cap on payments-per-person may be lower, but we try to make our process as fair and transparent as possible, we just need your understanding.

Thanks!
